ocaml/xapi-idl/cluster/cluster_cli.ml

```diff
@@ -8,23 +8,29 @@ let version_str description =
   let maj, min, mic = description.Idl.Interface.version in
   Printf.sprintf "%d.%d.%d" maj min mic
 
-let default_cmd =
-  let doc =
-    String.concat ""
-      [
-        "A CLI for the cluster API. This tool is not intended to be used as an "
-      ; "end user tool"
-      ]
-  in
-  ( Cmdliner.Term.(ret (const (fun _ -> `Help (`Pager, None)) $ const ()))
-  , Cmdliner.Term.info "cluster_cli"
-      ~version:(version_str Cmds.description)
-      ~doc
-  )
+open! Cmdliner
+
+let cmds =
+  List.map
+    (fun t ->
+      let t, i = t (Cluster_client.rpc_internal Cluster_client.json_url) in
+      Cmd.v i t
+    )
+    (Cmds.implementation ())
 
 let cli () =
-  let rpc = Cluster_client.rpc_internal Cluster_client.json_url in
-  Cmdliner.Term.eval_choice default_cmd
-    (List.map (fun t -> t rpc) (Cmds.implementation ()))
+  let default = Term.(ret (const (fun _ -> `Help (`Pager, None)) $ const ())) in
+  let info =
+    let doc =
+      String.concat ""
+        [
+          "A CLI for the cluster API. This tool is not intended to be used as "
+        ; "an end user tool"
+        ]
+    in
+    Cmd.info "cluster_cli" ~version:(version_str Cmds.description) ~doc
+  in
+  let cmd = Cmd.group ~default info cmds in
+  Cmd.eval_value cmd
 
-let _ = cli ()
+let () = match cli () with Ok (`Ok f) -> f () | _ -> ()
```

Comment:
Can't we use Cmd.eval_result or exit (...) in some way here to also ensure that errors get printed properly?

Reply:
We could yes, At this point it might be worth thinking about a function in xcp_service to reduce repetition here

Reply:
Yes something in xapi-idl/, all these CLIs are almost the same except for a few things like names, and those could easily be parameters. Perhaps do that as a commit on top of the other ones? (the conversion from old to new cmdliner is still good to see separately and then the refactoring to use common code).

Reply:
indeed